pkg/dogstatsd/listeners/packet_buffer.go

```diff
@@ -0,0 +1,61 @@
+package listeners
+
+import (
+	"sync"
+	"time"
+)
+
+type packetBuffer struct {
+	packets       Packets
+	flushTimer    *time.Ticker
+	bufferSize    uint
+	outputChannel chan Packets
+	closeChannel  chan struct{}
+	m             sync.Mutex
+}
+
+func newPacketBuffer(bufferSize uint, flushTimer time.Duration, outputChannel chan Packets) *packetBuffer {
+	pb := &packetBuffer{
+		bufferSize:    bufferSize,
+		flushTimer:    time.NewTicker(flushTimer),
+		outputChannel: outputChannel,
+		packets:       make(Packets, 0, bufferSize),
+		closeChannel:  make(chan struct{}),
+	}
+	go pb.flushLoop()
+	return pb
+}
+
+func (pb *packetBuffer) flushLoop() {
+	for {
+		select {
+		case <-pb.flushTimer.C:
+			pb.m.Lock()
+			pb.flush()
+			pb.m.Unlock()
+		case <-pb.closeChannel:
```

Comment:
You would probably want to flush if the channel is closed, no?

Reply:
This does not play well with the implementation since we don't control the lifecycle of the channel we flush to and writing to a closed channel makes us panic. 
In the current use case we should only exit on an agent exit and from what I see in the aggregator code we don't force flush on exit

Reply:
Not really needed I think: the aggregator flush to the fowarder every 10s for dogstatsd and the forwarder might not send the transaction right away. Since we don't want to block the agent on exit for a few seconds we won't have time to handle the data anyway.